nomos-libp2p/src/lib.rs

```diff
@@ -79,11 +182,34 @@ impl Swarm {
             .with_tokio()
             .with_quic()
             .with_dns()?
-            .with_behaviour(|_| Behaviour::new(peer_id, config.gossipsub_config.clone()).unwrap())?
+            .with_behaviour(|keypair| {
+                Behaviour::new(
+                    peer_id,
+                    config.gossipsub_config.clone(),
+                    config.kademlia_config.clone(),
+                    config.identify_config.clone(),
+                    keypair.public(),
+                )
+                .unwrap()
+            })?
             .with_swarm_config(|c| c.with_idle_connection_timeout(IDLE_CONN_TIMEOUT))
             .build();
 
-        swarm.listen_on(Self::multiaddr(config.host, config.port))?;
+        let listen_addr = Self::multiaddr(config.host, config.port);
+        swarm.listen_on(listen_addr.clone())?;
+
+        // if kademlia is enabled and is not in client mode then it is operating in a
+        // server mode
+        if let Some(kademlia_config) = &config.kademlia_config {
+            if !kademlia_config.client_mode {
+                // libp2p2-kad server mode is implicitly enabled
+                // by adding external addressess
+                // <https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/CHANGELOG.md#0440>
+                let external_addr = listen_addr.with(Protocol::P2p(peer_id));
+                swarm.add_external_address(external_addr.clone());
+                tracing::info!("Added external address: {}", external_addr);
+            }
+        }
```

Comment:
Uhm. I do not think we have to expose this. It should always be a client and a server. All nodes should participate the same. Does it makes sense?

Reply:
This was mentioned in the PoC phase, I believe and there was a todo comment to make it configurable. Basically client mode in kademlia means that node participate networks but is not providing query capabilities, so basically kind of a light node that queries other nodes but does not improve the network by itself. These are mentioned as good candidates for unrealiable devices. 

I am not saying we need this just wanted to clarify.
The default is a server mode, this code just adds an option to be in client_mode as well.

Reply:
> This was mentioned in the PoC phase, I believe and there was a todo comment to make it configurable. Basically client mode in kademlia means that node participate networks but is not providing query capabilities, so basically kind of a light node that queries other nodes but does not improve the network by itself. These are mentioned as good candidates for unrealiable devices.
> 
> I am not saying we need this just wanted to clarify. The default is a server mode, this code just adds an option to be in client_mode as well.

Ah, yes, I recall now about this. You are correct, thanks for the explanation.

Reply:
Should be fixed now.